### DISCUSSION TITLE
Generate paths based on file content inside public folder

### DISCUSSION THREAD

**Discussion by oskery** [vercel/next.js#22161] [Help] [answered]
I'm importing data from csv files inside the public folder to generate paths and acquire content for a static site:

pages/item/[id].js
```javascript
import { dsv } from 'd3'

export async function getStaticPaths() {
  const items = await dsv(';', `${server}/data/items.csv`)
  return {
    paths: items.map((x) => ({ params: { id: x.slug } })),
    fallback: true
  }
}

export async function getStaticProps(context) {
  const item = await dsv(
    ',',
    `${server}/data/items_details.csv`,
    (d) => {
      if (d.slug === context.params.id) return d
    }
  )
  return {
    props: {
      item: item ? item[0] : null
    }
  }
}
```

Since d3.dsv() (just like fetch()) requires an absolute path I created a file that returns the right path depending on dev/prod:

config.js
```javascript
const dev = process.env.NODE_ENV !== 'production'

export const server = dev
  ? 'http://localhost:3000'
  : 'https://path-to-production'
```

This works well in development, but will of course cause build error in production since the files are not yet hosted. Is there a solution for this apart from hosting the files elsewhere?

**Comment by michrome** [comment, marked as answer]
Disclaimer: I haven't used D3. `getStaticProps` runs server side, so there's no need to get the data via an HTTP/fetch call.

Instead, read the CSV directly from the filesystem into a string, then pass that string into D3 e.g. [`d3.csvParse("foo,bar\n1,2");`](https://github.com/d3/d3-dsv/blob/master/README.md#d3-dsv).

**Reply by oskery** [reply]
Thank you! So reading using the fs-module or is there some alternative better practice?

**Reply by oskery** [reply]
To answer my own question: yes. I found my answer [in the docs](https://nextjs.org/docs/basic-features/data-fetching#reading-files-use-processcwd), recommending usage of process.cwd():

```javascript
import path from 'path'
import { promises as fs } from 'fs'
import { dsvFormat } from 'd3'

export async function getStaticProps(context) {
  const filePath = path.join(process.cwd(), 'public/data/items_details.csv')
  const fileString = await fs.readFile(filePath, 'utf8')
  
  const item = csv.parse(fileString, (d) => {
    if (d.slug === context.params.id) return d
  })

  return {
    props: {
      item: item ? item[0] : null
    }
  }
}
```
